XbSymbolDatabase.c

```diff
@@ -247,6 +311,76 @@ extern uint32_t EmuLocateFunction(OOVPA *Oovpa, uint32_t lower, uint32_t upper)
     return 0;
 }
 
+void XbSymbolRegisterSymbol(OOVPATable* OovpaTable, uint32_t address, xb_symbol_register_t register_func) {
+    if (OovpaTable != (void*)0) {
+
+        OOVPA* Oovpa = OovpaTable->Oovpa;
+
+        // do we need to save the found address?
+        if (Oovpa->XRefSaveIndex != XRefNoSaveIndex) {
+            // If XRef is not found, save it then register once.
+            if (XRefDataBase[Oovpa->XRefSaveIndex] == 0) {
+                XRefDataBase[Oovpa->XRefSaveIndex] = address;
+                register_func((void*)0, OovpaTable->szFuncName, address, OovpaTable->Version);
+            }
+        } else {
+            register_func((void*)0, OovpaTable->szFuncName, address, OovpaTable->Version);
+        }
+    }
+}
+
+bool XbSymbolScan(void* xbeData, char* section_name, uint32_t lower_bound, uint32_t upper_bound, xb_symbol_register_t register_func) {
```

Comment:
I made a discovery, void* xbeData isn't needed as lower_bound and upper_bound arguments are only two things we need. Since these are the memory addresses.

Reply:
There's a difference between the buffer pointer which has to be scanned, and the virtual address where this buffer should reside.

These two can be identical, but in the case of an IDA Pro plugin, the buffer data won't be loaded at the actual image base address, hence the need of the two.

Finally, we either need the start of the buffer, or an ending address (either the end of the buffer, or the end of virtual memory).

If I may express my thoughts, I would pass on the arguments as : buffer + size + image_base

Reply:
Uncertainties like this are exactly the reason behind my prior proposal to pass in only one argument - a struct that we can modify more easily than the function arguments themselves.

A few questions/remarks on the arguments you listed :
* uint32_t xbe_address, // is this the address of the buffer to be scanned, or it's virtual base address?
* uint32_t xbe_size, // if this size applies to the argument above, then it's fine either way
* char* section_name, // why is this needed?
* uint32_t section_address, // why pass in a section address at all?
* uint32_t section_size, // why would you need the size of the section? 
* _symbol_register_t register_func // this callback might be simpler with only argument too

Reply:
* uint32_t xbe_address, // Address in memory, it can be anywhere in memory.
* uint32_t xbe_size, // Correct, this is part of "XBE_MAX_VA" usage.
* char* section_name, // We can only scan specific section for accuracy reasons. Please, don't argue about this.
* uint32_t section_address, // This is the "lower_bound".
* uint32_t section_size, // This is part of "upper_bound". Instead of re-adjust upper_bound the program itself, we can input what's already existed in xbe's header.
* _symbol_register_t register_func // Already has a simple function arguments contain everything we need now and the future. This is in lockdown.